src/nodes/NodeManager.ts

```diff
@@ -326,23 +379,131 @@ class NodeManager {
   }
 
   /**
-   * Sets a node in the NodeGraph
+   * Adds a node to the node graph. This assumes that you have already authenticated the node.
+   * Updates the node if the node already exists.
+   * @param nodeId - Id of the node we wish to add
+   * @param nodeAddress - Expected address of the node we want to add
+   * @param blocking - Flag for if the operation should block or utilize the async queue
+   * @param force - Flag for if we want to add the node without authenticating or if the bucket is full.
+   * This will drop the oldest node in favor of the new.
+   * @param timeout Connection timeout timeout
    */
+  @ready(new nodesErrors.ErrorNodeManagerNotRunning())
   public async setNode(
     nodeId: NodeId,
     nodeAddress: NodeAddress,
+    blocking: boolean = false,
+    force: boolean = false,
+    timeout?: number,
   ): Promise<void> {
-    return await this.nodeGraph.setNode(nodeId, nodeAddress);
+    // When adding a node we need to handle 3 cases
+    // 1. The node already exists. We need to update it's last updated field
+    // 2. The node doesn't exist and bucket has room.
+    //  We need to add the node to the bucket
+    // 3. The node doesn't exist and the bucket is full.
+    //  We need to ping the oldest node. If the ping succeeds we need to update
+    //  the lastUpdated of the oldest node and drop the new one. If the ping
+    //  fails we delete the old node and add in the new one.
+    const nodeData = await this.nodeGraph.getNode(nodeId);
+    // If this is a new entry, check the bucket limit
+    const [bucketIndex] = this.nodeGraph.bucketIndex(nodeId);
+    const count = await this.nodeGraph.getBucketMetaProp(bucketIndex, 'count');
+    if (nodeData != null || count < this.nodeGraph.nodeBucketLimit) {
+      // Either already exists or has room in the bucket
+      // We want to add or update the node
+      await this.nodeGraph.setNode(nodeId, nodeAddress);
+      // Updating the refreshBucket timer
+      this.refreshBucketUpdateDeadline(bucketIndex);
+    } else {
+      // We want to add a node but the bucket is full
+      // We need to ping the oldest node
+      if (force) {
+        // We just add the new node anyway without checking the old one
+        const oldNodeId = (
+          await this.nodeGraph.getOldestNode(bucketIndex, 1)
+        ).pop()!;
+        this.logger.debug(
+          `Force was set, removing ${nodesUtils.encodeNodeId(
+            oldNodeId,
+          )} and adding ${nodesUtils.encodeNodeId(nodeId)}`,
+        );
+        await this.nodeGraph.unsetNode(oldNodeId);
+        await this.nodeGraph.setNode(nodeId, nodeAddress);
+        // Updating the refreshBucket timer
+        this.refreshBucketUpdateDeadline(bucketIndex);
+        return;
+      }
+      if (blocking) {
```

Comment:
The logic here doesn't seem quite right to me - if `force` is set to true then, regardless of whether `blocking` is set or not, `nodeGraph.setNode()` is going to be called twice for the same node, right? Should it instead be `else if (blocking)`, or do we expect different behaviour if both `force` and `blocking` are set to true?

Reply:
A diagram for the different paths that can be taken when calling `NodeManager.setNode()` would be helpful here - has one been made?

Reply:
I don't have one atm. I believe there are 2 reasons for having a force vs block. Although this can be done in different ways.

1. Force means add the node to the node graph without checking if the node actually exists via a ping. This is often used when a ping has already occurred through one of the ways connections are made from the node or connections are made to the node. Or when the user explicitly wants to add the node for debugging purposes without caring about whether the node is alive.
2. Block is about choosing whether to run it by awaiting the result or by just putting it into the queue to run in the background. Sometimes you want the former and not the latter.

Reply:
Note that I have not reviewed this deeply. But this was all based on the fact that there are multiple ways a connection may come into existence and we had to handle all these ways elegantly.

Reply:
So it should be possible to specify both `force` and `blocking` and expect different behaviour? Because the current logic in `NodeManager.setNode()` is:

```
if (force):
  remove oldest node
  set new node immediately
if (blocking): // force can be true or false
  garbage collect old nodes
  set new node immediately if there's room // acts like force == false
else: // not blocking; force can be true or false
  queue set node
```

So if `force` is true then the node can be set twice. From your explanation it sounds like the logic should be more like:

```
if (blocking):
  garbage collect old nodes
  if (force && still no room):
    remove oldest node
  set new node immediately if there's room
else: // not blocking
  if (force):
    remove oldest node
    set new node immediately
  else:
    queue set node
```

Although in this case the `force` behaviour is essentially the same regardless of whether we're `blocking` or not, we're just performing extra (unnecessary) work in the former case. If we want to force a node to be set though then we always need to do that synchronously - if we add the operation to a queue then there's a chance that the bucket will be full again once we get around to it right? So I think just adding an `else` before `if (blocking)` is the most elegant solution.

```
if (force):
  force add synchronously
else if (blocking):
  garbage collect + add synchronously
else: // not blocking or force
  queue add asynchronously
```

Reply:
That could be correct. Just check if it makes sense in the larger context of the rest of the node adding/removal policies.

Reply:
From
> set a node, and if the bucket is full, drop the oldest node in the bucket (don't even use the concurrency parameter for the oldest) https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1082638907

and

> The setNode method has to deal with garbage collecting old nodes in a full bucket in case the force = false. https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1082760728

I believe that `else if (blocking)` is indeed the correct behaviour. I can fix that up.